Generate header with version #defines for LibTorch

jbschlosser opened the pull request:
Uses cmake's `configure_file()` macro to generate a new `torch/csrc/api/include/torch/version.h` header with `TORCH_VERSION_{MAJOR,MINOR,PATCH}` \#defines from an input file `torch/csrc/api/include/torch/version.h.in`.

For Bazel builds, this is accomplished with `genrule()` and a python generation script.

Fixes #44365

<img width="1229" alt="Screen Shot 2021-01-05 at 3 19 24 PM" src="https://user-images.githubusercontent.com/75754324/103809279-3fd80380-5027-11eb-9039-fd23922cebd5.png">

malfet commented on caffe2/CMakeLists.txt:
```diff
@@ -337,6 +337,10 @@ if(NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
     COPYONLY)
 
+  # Generate header with version info
+  configure_file("${TORCH_SRC_DIR}/csrc/api/include/torch/version.h.in"
```
You'll probably need to add `header_template_rule` to BUILD.bazel similar to 
https://github.com/pytorch/pytorch/blob/12ee7b61e75caa0c57b53ee1627e3075686d2510/BUILD.bazel#L41-L48

glaringlee replied:
> You'll probably need to add `header_template_rule` to BUILD.bazel similar to
> https://github.com/pytorch/pytorch/blob/12ee7b61e75caa0c57b53ee1627e3075686d2510/BUILD.bazel#L41-L48

@malfet 
It seems to me that we did not setup the version macros in BUILD.bazel (correct me if I am wrong)
I find this https://github.com/pytorch/pytorch/blob/master/BUILD.bazel#L786, but is it hard coded to me.

malfet replied:
@glaringlee we do not ship any binaries from bazel build, so version numbers can be stale.

glaringlee replied:
@malfet thanks for comment.
@jbschlosser per @malfet's comment above, it is fine that we just use a hard coded version for TORCH_VERSION_XXXX in bzl, but if you are in the middle of looking at bzl action to parse version.txt and it is not too time consuming, it is also great to let bzl support versioning. You can do either way, let me know when you think this is ready.

jbschlosser replied:
Great- thanks for looking into it! I have it working for now hard-coded, but I want to attempt the proper solution. Will let you know when I'm ready.

glaringlee replied:
sure, thank you!

jbschlosser replied:
@glaringlee Well, I now have the version header being generated from `version.txt` in Bazel. Since Bazel cannot read file contents at analysis time in a way that affects the build (see [here](https://stackoverflow.com/questions/56007307/read-local-file-content-in-bazel-build-system)), I could not use `header_template_rule` and had to switch to a `genrule` that launches a python script that does manual template expansion. It works but it's ugly. Perhaps it's best to keep the hard-coded version for now and solve this properly later on.

glaringlee replied:
@malfet This PR now uses `genrule` to parse version.txt in bazel build to support torch version number in c++.  The code introduced a helper .py script to parse version.txt and generate the version header file, it seems fine to me. Can you advise?

glaringlee commented:
If the bzl is not experimental and we need to setup the version correctly, given that bazel can not read file directly, we might need to create action to read version.txt and setup the TORCH_VERSION_XXXX just like what we do here, https://github.com/pytorch/pytorch/blob/9529ae377629f4253af0720cb01f09db6fcad4c9/CMakeLists.txt#L463

malfet commented on tools/setup_helpers/gen_version_header.py:
```diff
@@ -0,0 +1,67 @@
+            break
+
+    return tuple([int(n) for n in version_number_str.split(".")])
+
```
[PEP-8](https://www.python.org/dev/peps/pep-0008/#blank-lines) suggest 2 blank lines between functions

glaringlee replied:
@jbschlosser fix this, and we are good to land

malfet reviewed (approved):
Thank you for the thorough fix!

glaringlee reviewed (approved):
Approved, but need to fix the bazel test failure before landing.
I think we need to add the generated header here:
https://github.com/pytorch/pytorch/pull/50073/files#diff-7fc57714ef13c3325ce2a1130202edced92fcccc0c6db34a72f7b57f60d552a3R1903

glaringlee commented:
@jbschlosser 
It seems this passed the bazel build, but bazel test failed.

jbschlosser commented:
> Approved, but need to fix the bazel test failure before landing.
> I think we need to add the generated header here:
> https://github.com/pytorch/pytorch/pull/50073/files#diff-7fc57714ef13c3325ce2a1130202edced92fcccc0c6db34a72f7b57f60d552a3R1903

Ah, you're right. I thought the glob on `torch/csrc/**/*.h` would pick up the generated header, but it makes sense that it has to be explicitly included for the generation rule to trigger.

glaringlee reviewed (approved):
This is good now.

jbschlosser committed:
Generate header with version #defines for LibTorch (#50073)

Summary:
Uses cmake's `configure_file()` macro to generate a new `torch/csrc/api/include/torch/version.h` header with `TORCH_VERSION_{MAJOR,MINOR,PATCH}` \#defines from an input file `torch/csrc/api/include/torch/version.h.in`.

For Bazel builds, this is accomplished with `header_template_rule()`.

For Buck builds, this is accomplished with `fb_native.genrule()`.

Fixes https://github.com/pytorch/pytorch/issues/44365

<img width="1229" alt="Screen Shot 2021-01-05 at 3 19 24 PM" src="https://user-images.githubusercontent.com/75754324/103809279-3fd80380-5027-11eb-9039-fd23922cebd5.png">

Pull Request resolved: https://github.com/pytorch/pytorch/pull/50073

Reviewed By: glaringlee

Differential Revision: D25855877

Pulled By: jbschlosser

fbshipit-source-id: 56a0b19a4b0c989f4583fc3197e28d841a48f238
